### PR TITLE
Remove scikit-learn dependency

### PR DESCRIPTION
**Pull request by mohitcek**
This PR address the issue #3 and adds common classification metrics to the uqlm/utils directory.

Following metrics are included in this PR
- Accuracy
- Balanced Accuracy
- F1 Score
- F-beta Score
- Log-loss
- Precision
- Recall
- RocAUC Score

Tuner class and UQEnsemble demo notebook is updated to import required functions

Unit Tests
[] no new tests required
[x] new tests added
[] existing tests adjusted

### CODE REVIEW

File: examples/ensemble_tuning_demo.ipynb [hunk 1]
```diff
@@ -93,10 +93,9 @@
    "source": [
     "import os\n",
     "import numpy as np\n",
-    "from sklearn.metrics import precision_score, recall_score, f1_score\n",
     "\n",
     "from uqlm import UQEnsemble\n",
-    "from uqlm.utils import load_example_dataset, math_postprocessor, plot_model_accuracies"
+    "from uqlm.utils import load_example_dataset, math_postprocessor, plot_model_accuracies, precision_score, recall_score, f1_score"
```
Comment (dylanbouchard) [inline examples/ensemble_tuning_demo.ipynb:98]: Can we make this update in the other demos too?

**Review by dylanbouchard — CHANGES_REQUESTED** [review]
Can we please make the following updates:

1. Update import statements for other demos
2. Add screenshots to show matching values of these functions with corresponding `scikit-learn` functions?

**Comment by dylanbouchard** [comment]
@virenbajaj can you please review?

File: uqlm/utils/metrics.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+    # Calculate AUC using trapezoidal rule
+    return np.trapz(tpr, fpr)
+
+
+def _check_lists(list1: list, list2: list, list2_name:str = "y_pred") -> None:
+    if len(list1) != len(list2):
+        raise ValueError(f"y_true and {list2_name} must have the same length.")
+    
+    if not isinstance(list1, np.ndarray):
+        list1 = np.array(list1)
```
Comment (virenbajaj) [inline uqlm/utils/metrics.py:280]: I believe [np.asarray](https://numpy.org/doc/2.2/reference/generated/numpy.asarray.html) will do this in one line (only create a copy if its not a np.array)

File: uqlm/utils/metrics.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+
+    Returns
+    -------
+    float
+        Accuracy score or the number of correct predictions.
+    """
+    if check_inputs:
+        y_true, y_pred = _check_lists(list1=y_true, list2=y_pred)
+
+    return sum(y_true == y_pred)/len(y_true)
```
Comment (virenbajaj) [inline uqlm/utils/metrics.py:42]: using np operators like np.sum or np.count_nonzero on np arrays will be faster than python native operators like sum

https://stackoverflow.com/questions/10922231/pythons-sum-vs-numpys-numpy-sum#:~:text=Note%20that%20this%20is%20a,where%20Pythons%20sum%20is%20faster).

File: uqlm/utils/metrics.py [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+    -------
+    float
+        Balanced accuracy score.
+    """
+    if check_inputs:
+        y_true, y_pred = _check_lists(list1=y_true, list2=y_pred)
+    
+    classes = np.unique(y_true)
+    recall_scores = []
+    for class_ in classes:
```
Comment (virenbajaj) [inline uqlm/utils/metrics.py:70]: I believe this for loop makes this function non-vectorized

Comment (virenbajaj) [reply uqlm/utils/metrics.py:70]: For ref: [balanced accuracy sklearn source code](https://github.com/scikit-learn/scikit-learn/blob/98ed9dc73/sklearn/metrics/_classification.py#L2442) creates a confusion matrix first.

Comment (virenbajaj) [reply uqlm/utils/metrics.py:70]: LLM generated response for vectorizing (pls validate): 
```python
def balanced_accuracy_score(y_true: ArrayLike, y_pred: ArrayLike, check_inputs: bool=True, value_if_undefined: float=0.0) -> float:
    """
    Computes the balanced accuracy score.
    Parameters
    ----------
    y_true : ArrayLike
        A list or numpy.ndarray of True labels.
    y_pred : ArrayLike
        A list of Predicted labels.
    check_inputs : bool
        Whether to validate the input arrays.
    value_if_undefined : float
        Value to return if no actual positives are present in the sample.
    Returns
    -------
    float
        Balanced accuracy score.
    """
    if check_inputs:
        y_true, y_pred = _check_lists(list1=y_true, list2=y_pred)

    classes = np.unique(y_true)
    
    # Create binary masks for each class
    y_true_masks = y_true == classes[:, np.newaxis]  # Shape: (n_classes, n_samples)
    y_pred_masks = y_pred == classes[:, np.newaxis]  # Shape: (n_classes, n_samples)
    
    # Calculate true positives and actual positives for all classes at once
    true_positives = np.sum(y_true_masks * y_pred_masks, axis=1)
    actual_positives = np.sum(y_true_masks, axis=1)
    
    # Handle cases where there are no actual positives
    recall_scores = np.where(actual_positives == 0,
                           value_if_undefined,
                           true_positives / actual_positives)
    
    return np.mean(recall_scores)
```

**Review by virenbajaj — APPROVED** [review]
Left my comments mainly w.r.t vectorization (can incorporate them now or later - as you wish).

File: uqlm/utils/metrics.py (hunk 4 above)
Comment (mohitcek) [reply uqlm/utils/metrics.py:70]: Thank you for the suggestion, I left the loop for balanced accuracy because for our toolkit, it will be executed for two values (either [0, 1] or [True, False]). But, I will include suggested changes (after verify the solution) to be consistent.

**Review by virenbajaj — APPROVED** [review]
Thanks for making the changes :) Looks good to me!
